source/particles/particle_handler.cc

```diff
@@ -1845,76 +1833,195 @@ namespace Particles
+  {
+    register_store_callback();
+  }
+
+
+
+  template <int dim, int spacedim>
+  void
+  ParticleHandler<dim, spacedim>::register_store_callback()
   {
     parallel::distributed::Triangulation<dim, spacedim>
-      *non_const_triangulation =
+      *distributed_triangulation =
         const_cast<parallel::distributed::Triangulation<dim, spacedim> *>(
           dynamic_cast<const parallel::distributed::Triangulation<dim, spacedim>
                          *>(&(*triangulation)));
-    (void)non_const_triangulation;
+    (void)distributed_triangulation;
 
-    Assert(non_const_triangulation != nullptr, dealii::ExcNotImplemented());
+    Assert(
+      distributed_triangulation != nullptr,
+      ExcMessage(
+        "Mesh refinement in a non-distributed triangulation is not supported "
+        "by the ParticleHandler class. Either insert particles after mesh "
+        "creation and do not refine afterwards, or use a distributed triangulation."));
 
 #ifdef DEAL_II_WITH_P4EST
-    // Only save and load particles if there are any, we might get here for
-    // example if somebody created a ParticleHandler but generated 0
-    // particles.
-    update_cached_numbers();
+    const auto callback_function =
+      [this](
+        const typename Triangulation<dim, spacedim>::cell_iterator
+          &                                                     cell_iterator,
+        const typename Triangulation<dim, spacedim>::CellStatus cell_status) {
+        return this->store_particles(cell_iterator, cell_status);
+      };
+
+    handle = distributed_triangulation->register_data_attach(
+      callback_function, /*returns_variable_size_data=*/true);
+#endif
+  }
 
-    if (global_max_particles_per_cell > 0)
-      {
-        const auto callback_function =
-          [this](const typename Triangulation<dim, spacedim>::cell_iterator
-                   &cell_iterator,
-                 const typename Triangulation<dim, spacedim>::CellStatus
-                   cell_status) {
-            return this->store_particles(cell_iterator, cell_status);
-          };
 
-        handle = non_const_triangulation->register_data_attach(
-          callback_function, /*returns_variable_size_data=*/true);
-      }
-#endif
+
+  template <int dim, int spacedim>
+  void
+  ParticleHandler<dim, spacedim>::transfer_after_coarsening_and_refinement()
+  {
+    register_load_callback(false);
   }
 
 
 
+  template <int dim, int spacedim>
+  void
+  ParticleHandler<dim, spacedim>::deserialize()
+  {
+    register_load_callback(true);
+  }
+
+
   template <int dim, int spacedim>
   void
   ParticleHandler<dim, spacedim>::register_load_callback_function(
     const bool serialization)
+  {
+    if (serialization == true)
+      deserialize();
+    else
+      transfer_after_coarsening_and_refinement();
+  }
+
+
+
+  template <int dim, int spacedim>
+  void
+  ParticleHandler<dim, spacedim>::register_load_callback(
```

Comment:
I find this function name misleading. You don't register a callback function here, but actually perform the unpacking by calling `notify_ready_to_unpack`.

Maybe call it `{call|perform|trigger|activate}_load_callback` instead?